In [6]:
import pandas as pd
import numpy as np
df=pd.read_csv('adult.csv')
df

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,22,Private,310152,Some-college,10,Never-married,Protective-serv,Not-in-family,White,Male,0,0,40,United-States,<=50K
32557,27,Private,257302,Assoc-acdm,12,Married-civ-spouse,Tech-support,Wife,White,Female,0,0,38,United-States,<=50K
32558,40,Private,154374,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,40,United-States,>50K
32559,58,Private,151910,HS-grad,9,Widowed,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K


# Task 1

In [15]:
print(f"Data Set Information \n",df.info())
print(f"Head Columns \n \n",df.head())
count=df.isnull().sum()
print(f'Null Values count ',count)
df.replace('?',np.nan,inplace=True)
print(f'\n After replacement of ? \n',df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       30725 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      30718 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  31978 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB
Data Set Information 
 None
Head Columns 
 
    age workclass  fnlwgt     education  educatio

# Task 2

In [19]:
rows_count,colums_count=df.shape
print("Rows:",rows_count)
percentage=(count*100/rows_count)
print(percentage)
percentage.sort_values(ascending=False)

Rows: 32561
age               0.000000
workclass         5.638647
fnlwgt            0.000000
education         0.000000
education.num     0.000000
marital.status    0.000000
occupation        5.660146
relationship      0.000000
race              0.000000
sex               0.000000
capital.gain      0.000000
capital.loss      0.000000
hours.per.week    0.000000
native.country    1.790486
income            0.000000
dtype: float64


,0
occupation,5.660146
workclass,5.638647
native.country,1.790486
fnlwgt,0.000000
education,0.000000
education.num,0.000000
age,0.000000
marital.status,0.000000
relationship,0.000000
sex,0.000000


# Task 3

In [27]:
df['workclass'].replace(np.nan,"Unknown",inplace=True)
df['occupation'].replace(np.nan,"Unknown",inplace=True)
df.head()

/tmp/ipykernel_621/2875732382.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['workclass'].replace(np.nan,"Unknown",inplace=True)
/tmp/ipykernel_621/2875732382.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', 

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,Unknown,77053,HS-grad,9,Widowed,Unknown,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,Unknown,186061,Some-college,10,Widowed,Unknown,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


# Task 4

In [28]:
print(f'Duplicated Rows are : ',df.duplicated().sum())

Duplicated Rows are :  24


# Task 5

In [29]:
print(df['education'].unique())


['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']


In [30]:
print(df['marital.status'].unique())

['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']


In [31]:
print(df['native.country'].unique())

['United-States' nan 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan' 'India'
 'Philippines' 'Trinadad&Tobago' 'Canada' 'South' 'Holand-Netherlands'
 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany' 'Italy' 'Japan' 'Hong'
 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru' 'Nicaragua'
 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary' 'Columbia'
 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia' 'Scotland'
 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']


# Task 6

In [32]:
df.describe(include=[np.number])

,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [34]:
df.describe(include='object')

,workclass,education,marital.status,occupation,relationship,race,sex,native.country,income
count,32561,32561,32561,32561,32561,32561,32561,31978,32561
unique,9,16,7,15,6,5,2,41,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22696,10501,14976,4140,13193,27816,21790,29170,24720


# Task 7

In [36]:
df.value_counts('occupation')

,count
occupation,
Prof-specialty,4140
Craft-repair,4099
Exec-managerial,4066
Adm-clerical,3770
Sales,3650
Other-service,3295
Machine-op-inspct,2002
Unknown,1843
Transport-moving,1597


In [42]:
male_counts,female_counts=df.value_counts('sex')
print(f'Male Distribution:',male_counts*100/(male_counts+female_counts))
print(f'female_counts:',female_counts*100/(male_counts+female_counts))


Male Distribution: 66.92054912318417
female_counts: 33.07945087681582


In [43]:
income_less_50,income_more_50=df.value_counts('income')
print(f'income_less_50:',income_less_50*100/(income_less_50+income_more_50))
print(f'income_more_50:',income_more_50*100/(income_less_50+income_more_50))

income_less_50: 75.91904425539757
income_more_50: 24.080955744602438


# Task 8

In [45]:
df.groupby('education')['education.num'].unique()

,education.num
education,
10th,[6]
11th,[7]
12th,[8]
1st-4th,[2]
5th-6th,[3]
7th-8th,[4]
9th,[5]
Assoc-acdm,[12]
Assoc-voc,[11]


# Task 9

In [55]:
import sqlite3
conn=sqlite3.connect(':memory:')
df.to_sql('adult_income',conn,if_exists='replace',index=False)
sql_query="SELECT * FROM adult_income WHERE age > 30"
df_sql=pd.read_sql(sql_query,conn)
print(df_sql.shape)
df_filtered_original=df[df['age']>30]
print(df_filtered_original.shape)

if df_sql.shape==df_filtered_original.shape:
  print('Comparison Succesful')
else:
  print('Comparison Failed')
conn.close()

(21989, 15)
(21989, 15)
Comparison Succesful


# Task 10

1. there are 32561 rows and 15 cols
2. there are some missing values in the dataset , 24 duplicated rows
3.  Income class imbalance as we have calculated the percentage of income
  less than 50k is 75.92 % and greater than 50k is 24.08 %
4. Null values in the columns occupation and workclass are replaced by unknown.
They carried out 5% of data so therefore we have replaced it instead of dropping them.